src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -2870,34 +2959,55 @@ ss::future<bool> tx_gateway_frontend::try_create_tx_topic() {
 
 void tx_gateway_frontend::expire_old_txs() {
     ssx::spawn_with_gate(_gate, [this] {
-        auto shard = _shard_table.local().shard_for(model::legacy_tm_ntp);
-
-        if (shard == std::nullopt) {
-            rearm_expire_timer();
+        auto ntp_meta = _metadata_cache.local().get_topic_metadata(
+          model::tx_manager_nt);
+        if (!ntp_meta) {
+            vlog(
+              txlog.error,
+              "Topic {} doesn't exist in metadata cache,",
+              model::tx_manager_nt);
             return ss::now();
         }
+        return ss::do_for_each(
+          ntp_meta->get_assignments(), [this](partition_assignment pa) {
+              auto tx_ntp = model::ntp(
+                model::tx_manager_nt.ns, model::tx_manager_nt.tp, pa.id);
 
-        return container()
-          .invoke_on(
-            *shard,
-            _ssg,
-            [](tx_gateway_frontend& self) {
-                return ss::with_gate(self._gate, [&self] {
-                    return self.with_stm(
-                      [&self](checked<ss::shared_ptr<tm_stm>, tx_errc> r) {
-                          if (!r) {
-                              return ss::now();
-                          }
-                          auto stm = r.value();
-                          return stm->read_lock().then(
-                            [&self, stm](ss::basic_rwlock<>::holder unit) {
-                                return self.expire_old_txs(stm).finally(
-                                  [u = std::move(unit)] {});
-                            });
-                      });
-                });
-            })
-          .finally([this] { rearm_expire_timer(); });
+              auto shard = _shard_table.local().shard_for(tx_ntp);
+
+              if (shard == std::nullopt) {
+                  // TODO: Create per shard timer
+                  // https://github.com/redpanda-data/redpanda/issues/9606
+                  rearm_expire_timer();
+                  return ss::now();
+              }
+
+              return container()
```

Comment:
nit: again here: it's very hard to see what is being changed because the formatting is combined with the change.